File: best-time-to-buy.cc
#include <vector>
using namespace std;

class Solution {
public:
  int maxProfit(vector<int>& prices) {
    if (prices.empty()) return 0;
    int min_price = prices[0];
    int ans = 0;
    for (int i = 1; i < prices.size(); i++) {
      if (prices[i] - min_price > ans) {
        ans = prices[i] - min_price;
      }
      if (prices[i] < min_price) {
        min_price = prices[i];
      }
    }
    return ans;
  }
};

File: best-time-to-buy.java
public class Solution {
    public int maxProfit(int[] prices) {
        if (prices.length == 0) return 0;
        int minPrice = prices[0];
        int ans = 0;
        for (int i = 1; i < prices.length; i++) {
            ans = Math.max(ans, prices[i] - minPrice);
            minPrice = Math.min(minPrice, prices[i]);
        }
        return ans;
    }
}
